plugins/destination/clickhouse/client/migrate.go

```diff
@@ -12,67 +14,140 @@ import (
 
 // Migrate relies on the CLI/client to lock before running migration.
 func (c *Client) Migrate(ctx context.Context, tables schema.Tables) error {
-	current, err := c.getTableDefinitions(ctx)
+	currentSchema, err := c.getTableDefinitions(ctx)
 	if err != nil {
 		return err
 	}
 
+	newSchema := queries.NormalizedTables(tables)
+	if c.spec.MigrateMode != specs.MigrateModeForced {
+		nonSafeMigratableTables, changes := c.nonAutoMigrableTables(newSchema, currentSchema)
+		if len(nonSafeMigratableTables) > 0 {
+			return fmt.Errorf("tables %s with changes %v require force migration. use 'migrate_mode: forced'", strings.Join(nonSafeMigratableTables, ","), changes)
+		}
+	}
+
 	const maxConcurrentMigrate = 10
 	eg, ctx := errgroup.WithContext(ctx)
 	eg.SetLimit(maxConcurrentMigrate)
 
-	for _, table := range queries.NormalizedTables(tables) {
+	for _, table := range newSchema {
 		table := table
-		eg.Go(func() (err error) {
+		eg.Go(func() error {
 			c.logger.Info().Str("table", table.Name).Msg("Migrating table started")
-			defer func() {
-				c.logger.Err(err).Str("table", table.Name).Msg("Migrating table done")
-			}()
-
 			if len(table.Columns) == 0 {
 				c.logger.Warn().Str("table", table.Name).Msg("Table with no columns, skip")
 				return nil
 			}
 
-			curr := current.Get(table.Name)
-			if curr == nil {
+			current := currentSchema.Get(table.Name)
+			if current == nil {
 				return c.createTable(ctx, table)
 			}
 
-			return c.autoMigrateTable(ctx, curr, table)
+			err := c.autoMigrate(ctx, table, current)
+			if err != nil {
+				return err
+			}
+			c.logger.Err(err).Str("table", table.Name).Msg("Migrating table done")
+			return nil
 		})
 	}
 
 	return eg.Wait()
 }
 
+func (c *Client) nonAutoMigrableTables(tables schema.Tables, currentTables schema.Tables) ([]string, [][]schema.TableColumnChange) {
+	var result []string
+	var tableChanges [][]schema.TableColumnChange
+	for _, t := range tables {
+		current := currentTables.Get(t.Name)
+		if current == nil {
+			continue
+		}
+		changes := t.GetChanges(current)
+		if !c.canSafeMigrate(changes) {
+			result = append(result, t.Name)
+			tableChanges = append(tableChanges, changes)
+		}
+	}
+	return result, tableChanges
+}
+
+func (*Client) canSafeMigrate(changes []schema.TableColumnChange) bool {
+	for _, change := range changes {
+		if change.Type == schema.TableColumnChangeTypeAdd && change.Current.CreationOptions.NotNull {
+			return false
+		}
+
+		if change.Type == schema.TableColumnChangeTypeRemove && change.Previous.CreationOptions.NotNull {
+			return false
+		}
+
+		if change.Type == schema.TableColumnChangeTypeUpdate {
+			return false
+		}
```

Comment:
we have 2 options for `change.Current.CreationOptions.NotNull` to exit early (`schema.TableColumnChangeTypeAdd` & `schema.TableColumnChangeTypeRemove`).
Maybe we can go with this code instead?
```suggestion
		switch change.Type {
		case schema.TableColumnChangeTypeUpdate:
			return false
		case schema.TableColumnChangeTypeAdd,
			schema.TableColumnChangeTypeRemove:
			if change.Current.CreationOptions.NotNull {
				return false
			}
		}
```

Reply:
This suggestion is not the same as the previous code. `TableColumnChangeTypeRemove` looks at the previous column and `TableColumnChangeTypeAdd` looks at the current column.